text/0000-macro-metavar-expr.md

```diff
@@ -0,0 +1,332 @@
+                    ${ignore(x)}
+                )*
+            )*
+        )*
+    };
+}
+```
+
+Given this input:
+```
+    example! {
+        [ ( A B C D ) ( E F G H ) ( I J ) ]
+        [ ( K L M ) ]
+    }
+```
+
+The macro would expand to:
+```
+    counts = (2, 4, 13)
+    nested:
+        indexes = (0/2)
+        counts = (3, 10)
+        nested:
+            indexes = (0/2, 0/3)
+            counts = (4)
+            nested:
+                indexes = (0/2, 0/3, 0/4)
+                indexes = (0/2, 0/3, 1/4)
+                indexes = (0/2, 0/3, 2/4)
+                indexes = (0/2, 0/3, 3/4)
+            indexes = (0/2, 1/3)
+            counts = (4)
+            nested:
+                indexes = (0/2, 1/3, 0/4)
+                indexes = (0/2, 1/3, 1/4)
+                indexes = (0/2, 1/3, 2/4)
+                indexes = (0/2, 1/3, 3/4)
+            indexes = (0/2, 2/3)
+            counts = (2)
+            nested:
+                indexes = (0/2, 2/3, 0/2)
+                indexes = (0/2, 2/3, 1/2)
+        indexes = (1/2)
+        counts = (1, 3)
+        nested:
+            indexes = (1/2, 0/1)
+            counts = (3)
+            nested:
+                indexes = (1/2, 0/1, 0/3)
+                indexes = (1/2, 0/1, 1/3)
+                indexes = (1/2, 0/1, 2/3)
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds additional syntax to the language, that program authors must learn
+and understand.  We may not want to add more syntax.
+
+The author believes it is worth the overhead of new syntax, as even though
+there exist workarounds for obtaining the information if it's really needed,
+these workarounds are sometimes difficult to discover and naive
+implementations can significantly harm compiler performance.
+
+Furthermore, the additional syntax is limited to declarative macros, and its
+use should be limited to specific circustances where it is more understandable
+than the alternatives.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This RFC proposes a modest but powerful extension to macro syntax that makes
+it possible to obtain information that the compiler already knows, but
+requires inefficient and complex techniques to obtain in the macro.
+
+The original proposal was for a shorter syntax to provide the count of
+repetitions: `$#ident`.  During discussions of this syntax, it became clear
+that it was not obvious as to which number this referred to: the count of
+repetitions at this level, or the length of the current repetition.  It also
+does not provide a way to discover counts or lengths for other repetition
+depths.  There was also interest in being able to discover the index of the
+current repetition, and the `#` character had been used in similar proposals
+for that.  There was some reservation expressed for the use of the `#` token
+because of the cognitive burden of another sigil, and its common use in the
+`quote!` macro.
```

Comment:
IIRC, I already asked this on Zulip, but could you document what syntactic space is available for this feature (and similar features) in general?

The main issue with choosing a syntax here is that pretty much any syntax is valid on the right hand side of the macro because it represents macro output, which can be an arbitrary token stream.
So, we are lucky that the combination `$ {` turned out reserved.
Are any other similar reserved combinations that can be used for new macro features?

Reply:
I believe `$[ ... ]` was also available

Reply:
As @Lokathor says, this is mentioned briefly in the "Future possibilities" section at the bottom, but I will expand on it.

Currently the `$` symbol can only be followed by an identifier or `(`, so anything else can be used in a language extension.  This RFC specifies `${ .. }` and `$$`, but `$[ .. ]` remains invalid, as does `$` followed by any other symbol (so `$@`, `$:`, `$!` or similar could be used).

Additionally, metavariable expressions are intended to be extensible themselves.  This RFC defines `count`, `index`, `length` and `ignore`, but future RFCs can add additional expressions of the form `${foo(fooargs...)}`.  Anything that fits within this pattern and can be evaluated by the macro transcriber would be a suitable candidate for another expression.

Reply:
I like this, and I like that it carves out an extensible space for future improvements to macro syntax.

Reply:
I worry slightly that `${...}` and `$(...)` might look too similar. That might be an artifact of my font.

In particular, the its going to be tricky to give good diagnostics when a user writes `$(...)` when they meant to write `${...}`, and vice versa. Especially if their macro body happens to refer to names like `count` or `length`

But I don't have great counter-suggestions; `$[...]` might be just as bad (though I *do* think it is easier to distinguish from `$(...)`.) The only other counter-suggestion  can think of is `${{...}}`, but that might be  bridge too far.

Reply:
I see your worry.  I also have some reservations about the `)}` cluster in `${count(ident)}`.

We don't have to use a delimited block.  Since currently anything other than `ident` or `(..)` is invalid after `$` we could use some other sigil.  Some examples:
* `$:count(ident)`  e.g. `let v = Vec::with_capacity($:count(values))`
* `$@count(ident)`  e.g. `let v = Vec::with_capacity($@count(values))`
* `$!count(ident)` e.g. `let v = Vec::with_capacity($!count(values))`

Using the last one as an example, this would be parsed as: `$` `!` `<metavar-expr-function>` `(` `<metavar-expr-args...>` `)`

Other suggestions also welcome.